packages/x-date-pickers/src/internals/hooks/useField/useField.utils.ts

```diff
@@ -617,15 +734,19 @@ export const getDateSectionGetterAndSetter = <TDate>(
 
   return adapterMethods[dateSectionName];
 };
-export const cleanTrailingZeroInNumericSectionValue = (value: string, maximum: number) => {
```

Comment:
Date-fns has a format with a trailing zero on numeral week days (`01-07`).
But `7.toString().length === 1`, so our current method does not work (my assumption was that the maximum boundary had no trailing zero).

The new logic is simply to format the current date with the format and check the length.
The assumption is now that all values have the same length on a section with trailing zeroes (no `002` but `0204` for example).
It's a looser assumption, I hope this one will hold.

Reply:
It does look and sound that this solution should be more robust 👍